### PR TITLE
Fix incremental recovery failure because the checkpoint redo wal file before divergence LSN is gone.

### PR DESCRIPTION
**Pull request by paul-guo-**
This is the master version of https://github.com/greenplum-db/gpdb/pull/11891

Previously we thought the fix for master and that for gp6 are different so created a gp6 version at first. Now we found the master version is similar, so port the gp6 version. The code change is a bit different but the theory is the same.

### CODE REVIEW

File: src/backend/access/transam/xlog.c [hunk 1]
```diff
@@ -9231,6 +9237,9 @@ CreateCheckPoint(int flags)
 	 */
 	PriorRedoPtr = ControlFile->checkPointCopy.redo;
 
+	/* Do not use PriorRedoPtr directly to reduce future merge conflict.*/
+	ControlFileOldCheckpointCopyRedo = PriorRedoPtr;
```
Comment (ashwinstar) [inline src/backend/access/transam/xlog.c:9241]: Based on above comment `PriorRedoPtr` can be passed to `KeepLogSeg()` once interface is modified.

**Review by ashwinstar — APPROVED** [review]
:+1: LGTM

File: src/backend/access/transam/xlog.c [hunk 2]
```diff
@@ -881,6 +881,12 @@ static bool holdingAllLocks = false;
 static MemoryContext walDebugCxt = NULL;
 #endif
 
+/*
+ * gpdb: backup of ControlFile->checkPointCopy.redo.  This is used by
+ * KeepLogSeg() to avoid pg_rewind failure due to missing xlog file.
+ */
+static XLogRecPtr ControlFileOldCheckpointCopyRedo = InvalidXLogRecPtr;
```
[the remark this answers is not in the document]
Comment (paul-guo-) [reply src/backend/access/transam/xlog.c]: Yes thought about this solution when working on the PR also. Seems both solutions are all easy for merge conflict fixing in the future. OK, let's use `PriorRedoPtr` instead.

File: src/backend/access/transam/xlog.c [hunk 3]
```diff
@@ -9621,7 +9621,7 @@ CreateRestartPoint(int flags)
 	receivePtr = GetWalRcvWriteRecPtr(NULL, NULL);
 	replayPtr = GetXLogReplayRecPtr(&replayTLI);
 	endptr = (receivePtr < replayPtr) ? replayPtr : receivePtr;
-	KeepLogSeg(endptr, &_logSegNo);
+	KeepLogSeg(endptr, &_logSegNo, PriorRedoPtr);
```
Comment (ashwinstar) [inline src/backend/access/transam/xlog.c:9624]: I thought for RestartPoints we don't wish to trigger the logic hence will be passing `InvalidXLogRecPtr` here?

Comment (paul-guo-) [reply src/backend/access/transam/xlog.c]: Modified. The argument is useless in KeepLogSeg() for the restartpoint case though.

**Pull request merged**